File: typescript/oop-patterns/behavioral-patterns/strategy/strategy.ts
// Контекст используется клиентом для вызова методов, решающих бизнес
// задачи
class StrategyContext {
  private strategy: iStrategy;

  constructor(strategy: iStrategy) {
    this.strategy = strategy;
  }

  public setStrategy(strategy: iStrategy) {
    this.strategy = strategy;
  }

  // Контекст передаёт выполнение задачи в конкретный объект стратегии, при
  // этом не имея методов для её непосредственного решения внутри себя
  public doSomeBusinessLogic(): void {
    console.log('Контекст: Сортирую данные, используя стратегию');
    const result = this.strategy.doAlgorithm(['a', 'b', 'c', 'e', 'd']);
    console.log(result.join(','));
  }
}

// Интерфейс стратегии должен включать все поддерживаемые каждой стратегией
// методы, контекст стратегий использует этот интерфейс для вызова
// соответствующих методов на объекте стратегии
interface iStrategy {
  doAlgorithm(data: string[]): string[];
}

// Конкретные стратегии непосредственно включают реализацию интерфейса стратегий
class ConcreteStrategyA implements iStrategy {
  public doAlgorithm(data: string[]): string[] {
    return data.sort();
  }
}

class ConcreteStrategyB implements iStrategy {
  public doAlgorithm(data: string[]): string[] {
    return data.sort((a, b) => (a < b ? 1 : a > b ? -1 : 0));
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

const strategyContext = new StrategyContext(new ConcreteStrategyA());
console.log('Клиент: Установлена стратегия сортировки по возрастанию');
strategyContext.doSomeBusinessLogic();

console.log('');

console.log('Клиент: Установлена стратегия сортировки по убыванию');
strategyContext.setStrategy(new ConcreteStrategyB());
strategyContext.doSomeBusinessLogic();

File: go/oop-patterns/behavioral-patterns/strategy/strategy.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// Контекст используется клиентом для вызова методов, решающих бизнес
// задачи
type strategyContext struct {
	strategy iStrategy
}

func (c *strategyContext) setStrategy(strategy iStrategy) {
	c.strategy = strategy
}

// Контекст передаёт выполнение задачи в конкретный объект стратегии, при
// этом не имея методов для её непосредственного решения внутри себя
func (c *strategyContext) doSomeBusinessLogic() {

	fmt.Println("Контекст: Сортирую данные, используя стратегию")
	result := c.strategy.doAlgorithm([]string{"a", "b", "c", "d", "e"})
	fmt.Println(strings.Join(result, ","))
}

// Интерфейс стратегии должен включать все поддерживаемые каждой стратегией
// методы, контекст стратегий использует этот интерфейс для вызова
// соответствующих методов на объекте стратегии
type iStrategy interface {
	doAlgorithm(data []string) []string
}

// Конкретные стратегии непосредственно включают реализацию интерфейса стратегий
type concreteStrategyA struct {
}

func (s *concreteStrategyA) doAlgorithm(data []string) []string {
	sort.Strings(data)
	return data
}

type concreteStrategyB struct {
}

func (s *concreteStrategyB) doAlgorithm(data []string) []string {
	sort.Slice(data, func(i, j int) bool {
		return data[i] > data[j]
	})
	return data
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func main() {
	context := strategyContext{strategy: &concreteStrategyA{}}
	fmt.Println("Клиент: Установлена стратегия сортировки по возрастанию")
	context.doSomeBusinessLogic()

	fmt.Println("")

	fmt.Println("Клиент: Установлена стратегия сортировки по убыванию")
	context.setStrategy(&concreteStrategyB{})
	context.doSomeBusinessLogic()
}
